{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "m01 = [ 35, 43, 51, 59, 67, 75, 83, 91 ]\n",
    "m02 = [ 36, 44, 52, 60, 68, 76, 84, 92 ]\n",
    "m03 = [ 37, 45, 53, 61, 69, 77, 85, 93 ]\n",
    "m04 = [ 38, 46, 54, 62, 70, 78, 86, 94 ]\n",
    "m05 = [ 39, 47, 55, 63, 71, 79, 87, 95 ]\n",
    "m06 = [ 40, 48, 56, 64, 72, 80, 88, 96 ]\n",
    "m07 = [ 41, 49, 57, 65, 73, 81, 89, 97 ]\n",
    "m08 = [ 42, 50, 58, 66, 74, 82, 90, 98 ]\n",
    "m09 = [ 99, 101, 103, 105, 107, 109, 111, 113 ]\n",
    "m10 = [ 100, 102, 104, 106, 108, 110, 112, 114 ]\n",
    "\n",
    "allms = [ m01, m02, m03, m04, m05, m06, m07, m08, m09, m10 ]\n",
    "g_num_list = list( range( 1, 8 ) ) + list( range( 9, 25 ) ) + list( range( 26, 35 ) )\n",
    "\n",
    "# for tm00 in allms:\n",
    "#     print( 'Processing microcosm ' + str( tm00 ) + '. . .\\n' )\n",
    "#     for g_num in g_num_list:\n",
    "#         print( 'Processing genome ' + str( g_num ) + '. . .\\n' )\n",
    "#         snp_detects = {}\n",
    "#         is_invalid = False\n",
    "#         for this_idx, this_time in enumerate( tm00 ):\n",
    "#             # Going from time point to time point.\n",
    "#             try:\n",
    "#                 df = pd.read_csv( '/home/akshitg/lte/metagenomes_lte/freebayes_out/snps/fb_processed/mg_' \n",
    "#                                   + str( this_time ) + '_gnm_' + str( g_num ) + \n",
    "#                                   '_processed_snps.tsv', sep='\\t', index_col=0 )\n",
    "                \n",
    "#                 if len( df ) == 0:\n",
    "#                     is_invalid = True\n",
    "#                     continue\n",
    "#             except:\n",
    "#                 is_invalid = True\n",
    "#                 continue\n",
    "\n",
    "#             # Now go SNP by SNP.\n",
    "#             for idx, row in  df.iterrows():\n",
    "#                 tsnp = str( row[ 'contig' ] ) + '_' + str( row[ 'gene_num' ] ) + '_' + str( row[ 'pos' ] )\n",
    "#                 if tsnp in snp_detects:\n",
    "#                     snp_detects[ tsnp ] += 1\n",
    "#                 else:\n",
    "#                     snp_detects[ tsnp ] = 1\n",
    "\n",
    "#         if is_invalid:\n",
    "#             continue\n",
    "                        \n",
    "#         THRESHOLD_N = 3\n",
    "#         bad_snps = []\n",
    "#         for tsnp in snp_detects:\n",
    "#             if snp_detects[ tsnp ] < THRESHOLD_N:\n",
    "#                 bad_snps.append( tsnp )\n",
    "\n",
    "#         for tsnp in bad_snps:\n",
    "#             snp_detects.pop( tsnp, None )\n",
    "\n",
    "\n",
    "#         for this_idx, this_time in enumerate( tm00 ):\n",
    "#             try:\n",
    "#                 # Going from time point to time point.\n",
    "#                 df = pd.read_csv( '/home/akshitg/lte/metagenomes_lte/freebayes_out/snps/fb_processed/mg_' \n",
    "#                                   + str( this_time ) + '_gnm_' + str( g_num ) + \n",
    "#                                   '_processed_snps.tsv', sep='\\t', index_col=0 )\n",
    "                \n",
    "#                 if len( df ) == 0:\n",
    "#                     is_invalid = True\n",
    "#                     continue\n",
    "#             except:\n",
    "#                 is_invalid = True\n",
    "#                 continue\n",
    "\n",
    "#             ndf = pd.DataFrame()\n",
    "\n",
    "#             # Now go SNP by SNP.\n",
    "#             for idx, row in  df.iterrows():\n",
    "#                 tsnp = str( row[ 'contig' ] ) + '_' + str( row[ 'gene_num' ] ) + '_' + str( row[ 'pos' ] )\n",
    "#                 if tsnp in snp_detects:\n",
    "#                     ndf = ndf.append( df.iloc[ idx ] )\n",
    "#             ndf = ndf[ df.columns.tolist() ]\n",
    "#             ndf.to_csv( '/home/akshitg/lte/metagenomes_lte/freebayes_out/snps/fb_final/mg_' \n",
    "#                         + str( this_time ) + '_gnm_' + str( g_num ) + '_final_snps.tsv', sep='\\t' )\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing genome 8. . .\n",
      "\n"
     ]
    }
   ],
   "source": [
    "this_idx = 0\n",
    "this_time = m01[ this_idx ]\n",
    "tm00 = m01[:]\n",
    "g_num = 8\n",
    "print( 'Processing genome ' + str( g_num ) + '. . .\\n' )\n",
    "snp_detects = {}\n",
    "is_invalid = False\n",
    "for this_idx, this_time in enumerate( tm00 ):\n",
    "    # Going from time point to time point.\n",
    "    try:\n",
    "        df = pd.read_csv( '/home/akshitg/lte/metagenomes_lte/freebayes_out/snps/fb_processed/mg_' \n",
    "                          + str( this_time ) + '_gnm_' + str( g_num ) + \n",
    "                          '_processed_snps.tsv', sep='\\t', index_col=0 )\n",
    "\n",
    "        if len( df ) == 0:\n",
    "            is_invalid = True\n",
    "            continue\n",
    "    except:\n",
    "        is_invalid = True\n",
    "        continue\n",
    "\n",
    "    # Now go SNP by SNP.\n",
    "    for idx, row in  df.iterrows():\n",
    "        tsnp = str( row[ 'contig' ] ) + '_' + str( row[ 'gene_num' ] ) + '_' + str( row[ 'pos' ] )\n",
    "        if tsnp in snp_detects:\n",
    "            snp_detects[ tsnp ] += 1\n",
    "        else:\n",
    "            snp_detects[ tsnp ] = 1\n",
    "\n",
    "THRESHOLD_N = 3\n",
    "bad_snps = []\n",
    "for tsnp in snp_detects:\n",
    "    if snp_detects[ tsnp ] < THRESHOLD_N:\n",
    "        bad_snps.append( tsnp )\n",
    "\n",
    "for tsnp in bad_snps:\n",
    "    snp_detects.pop( tsnp, None )\n",
    "\n",
    "\n",
    "for this_idx, this_time in enumerate( tm00 ):\n",
    "    try:\n",
    "        # Going from time point to time point.\n",
    "        df = pd.read_csv( '/home/akshitg/lte/metagenomes_lte/freebayes_out/snps/fb_processed/mg_' \n",
    "                          + str( this_time ) + '_gnm_' + str( g_num ) + \n",
    "                          '_processed_snps.tsv', sep='\\t', index_col=0 )\n",
    "\n",
    "        if len( df ) == 0:\n",
    "            is_invalid = True\n",
    "            continue\n",
    "    except:\n",
    "        is_invalid = True\n",
    "        continue\n",
    "\n",
    "    ndf = pd.DataFrame()\n",
    "\n",
    "    # Now go SNP by SNP.\n",
    "    for idx, row in  df.iterrows():\n",
    "        tsnp = str( row[ 'contig' ] ) + '_' + str( row[ 'gene_num' ] ) + '_' + str( row[ 'pos' ] )\n",
    "        if tsnp in snp_detects:\n",
    "            ndf = ndf.append( df.iloc[ idx ] )\n",
    "    if len( ndf ):\n",
    "        ndf = ndf[ df.columns.tolist() ]\n",
    "#         ndf.to_csv( '/home/akshitg/lte/metagenomes_lte/freebayes_out/snps/fb_final/mg_' \n",
    "#                     + str( this_time ) + '_gnm_' + str( g_num ) + '_final_snps.tsv', sep='\\t' )\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>contig</th>\n",
       "      <th>gene_num</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref_base</th>\n",
       "      <th>alt_base</th>\n",
       "      <th>ref_aa</th>\n",
       "      <th>alt_aa</th>\n",
       "      <th>snp_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GNM_260008_NODE_66_length_37995_cov_12.771232</td>\n",
       "      <td>4224.0</td>\n",
       "      <td>13346.0</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>V</td>\n",
       "      <td>L</td>\n",
       "      <td>non</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GNM_260008_NODE_66_length_37995_cov_12.771232</td>\n",
       "      <td>4224.0</td>\n",
       "      <td>13391.0</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>syn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GNM_260008_NODE_66_length_37995_cov_12.771232</td>\n",
       "      <td>4224.0</td>\n",
       "      <td>13396.0</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>S</td>\n",
       "      <td>syn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GNM_260008_NODE_95_length_27924_cov_11.733101</td>\n",
       "      <td>5127.0</td>\n",
       "      <td>11898.0</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>K</td>\n",
       "      <td>K</td>\n",
       "      <td>syn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GNM_260008_NODE_95_length_27924_cov_11.733101</td>\n",
       "      <td>5127.0</td>\n",
       "      <td>11914.0</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>L</td>\n",
       "      <td>L</td>\n",
       "      <td>syn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>GNM_260008_NODE_95_length_27924_cov_11.733101</td>\n",
       "      <td>5127.0</td>\n",
       "      <td>11937.0</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>D</td>\n",
       "      <td>D</td>\n",
       "      <td>syn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>GNM_260008_NODE_95_length_27924_cov_11.733101</td>\n",
       "      <td>5127.0</td>\n",
       "      <td>11971.0</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>L</td>\n",
       "      <td>L</td>\n",
       "      <td>syn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>GNM_260008_NODE_95_length_27924_cov_11.733101</td>\n",
       "      <td>5127.0</td>\n",
       "      <td>12024.0</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>syn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>GNM_260008_NODE_95_length_27924_cov_11.733101</td>\n",
       "      <td>5127.0</td>\n",
       "      <td>12039.0</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>syn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>GNM_260008_NODE_95_length_27924_cov_11.733101</td>\n",
       "      <td>5127.0</td>\n",
       "      <td>12047.0</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>K</td>\n",
       "      <td>non</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>GNM_260008_NODE_268_length_2342_cov_22.624379</td>\n",
       "      <td>7204.0</td>\n",
       "      <td>1919.0</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>L</td>\n",
       "      <td>L</td>\n",
       "      <td>syn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>GNM_260008_NODE_268_length_2342_cov_22.624379</td>\n",
       "      <td>7204.0</td>\n",
       "      <td>1931.0</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>H</td>\n",
       "      <td>D</td>\n",
       "      <td>non</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>GNM_260008_NODE_268_length_2342_cov_22.624379</td>\n",
       "      <td>7204.0</td>\n",
       "      <td>1973.0</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>L</td>\n",
       "      <td>M</td>\n",
       "      <td>non</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>GNM_260008_NODE_1540_length_307_cov_27.555556</td>\n",
       "      <td>8572.0</td>\n",
       "      <td>127.0</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>non</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>GNM_260008_NODE_1553_length_307_cov_12.038889</td>\n",
       "      <td>8592.0</td>\n",
       "      <td>180.0</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>F</td>\n",
       "      <td>C</td>\n",
       "      <td>non</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                           contig  gene_num      pos ref_base  \\\n",
       "0   GNM_260008_NODE_66_length_37995_cov_12.771232    4224.0  13346.0        G   \n",
       "1   GNM_260008_NODE_66_length_37995_cov_12.771232    4224.0  13391.0        A   \n",
       "2   GNM_260008_NODE_66_length_37995_cov_12.771232    4224.0  13396.0        C   \n",
       "3   GNM_260008_NODE_95_length_27924_cov_11.733101    5127.0  11898.0        A   \n",
       "4   GNM_260008_NODE_95_length_27924_cov_11.733101    5127.0  11914.0        T   \n",
       "5   GNM_260008_NODE_95_length_27924_cov_11.733101    5127.0  11937.0        C   \n",
       "6   GNM_260008_NODE_95_length_27924_cov_11.733101    5127.0  11971.0        T   \n",
       "7   GNM_260008_NODE_95_length_27924_cov_11.733101    5127.0  12024.0        G   \n",
       "8   GNM_260008_NODE_95_length_27924_cov_11.733101    5127.0  12039.0        C   \n",
       "9   GNM_260008_NODE_95_length_27924_cov_11.733101    5127.0  12047.0        C   \n",
       "10  GNM_260008_NODE_268_length_2342_cov_22.624379    7204.0   1919.0        A   \n",
       "11  GNM_260008_NODE_268_length_2342_cov_22.624379    7204.0   1931.0        A   \n",
       "12  GNM_260008_NODE_268_length_2342_cov_22.624379    7204.0   1973.0        G   \n",
       "13  GNM_260008_NODE_1540_length_307_cov_27.555556    8572.0    127.0        C   \n",
       "14  GNM_260008_NODE_1553_length_307_cov_12.038889    8592.0    180.0        T   \n",
       "\n",
       "   alt_base ref_aa alt_aa snp_type  \n",
       "0         T      V      L      non  \n",
       "1         G      G      G      syn  \n",
       "2         T      S      S      syn  \n",
       "3         G      K      K      syn  \n",
       "4         C      L      L      syn  \n",
       "5         T      D      D      syn  \n",
       "6         C      L      L      syn  \n",
       "7         T      A      A      syn  \n",
       "8         G      T      T      syn  \n",
       "9         A      T      K      non  \n",
       "10        C      L      L      syn  \n",
       "11        G      H      D      non  \n",
       "12        A      L      M      non  \n",
       "13        G      C      G      non  \n",
       "14        G      F      C      non  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ndf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>contig</th>\n",
       "      <th>gene_num</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref_base</th>\n",
       "      <th>alt_base</th>\n",
       "      <th>ref_aa</th>\n",
       "      <th>alt_aa</th>\n",
       "      <th>snp_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GNM_260007_NODE_1_length_200824_cov_16.675431</td>\n",
       "      <td>99</td>\n",
       "      <td>106551</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>I</td>\n",
       "      <td>L</td>\n",
       "      <td>non</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GNM_260007_NODE_1_length_200824_cov_16.675431</td>\n",
       "      <td>99</td>\n",
       "      <td>106575</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>P</td>\n",
       "      <td>non</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GNM_260007_NODE_1_length_200824_cov_16.675431</td>\n",
       "      <td>99</td>\n",
       "      <td>106608</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>V</td>\n",
       "      <td>L</td>\n",
       "      <td>non</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GNM_260007_NODE_1_length_200824_cov_16.675431</td>\n",
       "      <td>99</td>\n",
       "      <td>106614</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>M</td>\n",
       "      <td>L</td>\n",
       "      <td>non</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GNM_260007_NODE_1_length_200824_cov_16.675431</td>\n",
       "      <td>99</td>\n",
       "      <td>106623</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>K</td>\n",
       "      <td>E</td>\n",
       "      <td>non</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>GNM_260007_NODE_110_length_10997_cov_22.147470</td>\n",
       "      <td>4768</td>\n",
       "      <td>6887</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>F</td>\n",
       "      <td>I</td>\n",
       "      <td>non</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84</th>\n",
       "      <td>GNM_260007_NODE_110_length_10997_cov_22.147470</td>\n",
       "      <td>4768</td>\n",
       "      <td>6896</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>syn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>GNM_260007_NODE_110_length_10997_cov_22.147470</td>\n",
       "      <td>4768</td>\n",
       "      <td>6938</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>R</td>\n",
       "      <td>R</td>\n",
       "      <td>syn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>GNM_260007_NODE_161_length_2736_cov_24.988118</td>\n",
       "      <td>5103</td>\n",
       "      <td>2007</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>syn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>GNM_260007_NODE_464_length_302_cov_117.988571</td>\n",
       "      <td>5547</td>\n",
       "      <td>175</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>S</td>\n",
       "      <td>non</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>88 rows Ã— 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                            contig  gene_num     pos ref_base  \\\n",
       "0    GNM_260007_NODE_1_length_200824_cov_16.675431        99  106551        G   \n",
       "1    GNM_260007_NODE_1_length_200824_cov_16.675431        99  106575        T   \n",
       "2    GNM_260007_NODE_1_length_200824_cov_16.675431        99  106608        A   \n",
       "3    GNM_260007_NODE_1_length_200824_cov_16.675431        99  106614        C   \n",
       "4    GNM_260007_NODE_1_length_200824_cov_16.675431        99  106623        T   \n",
       "..                                             ...       ...     ...      ...   \n",
       "83  GNM_260007_NODE_110_length_10997_cov_22.147470      4768    6887        G   \n",
       "84  GNM_260007_NODE_110_length_10997_cov_22.147470      4768    6896        A   \n",
       "85  GNM_260007_NODE_110_length_10997_cov_22.147470      4768    6938        G   \n",
       "86   GNM_260007_NODE_161_length_2736_cov_24.988118      5103    2007        T   \n",
       "87   GNM_260007_NODE_464_length_302_cov_117.988571      5547     175        G   \n",
       "\n",
       "   alt_base ref_aa alt_aa snp_type  \n",
       "0         C      I      L      non  \n",
       "1         C      T      P      non  \n",
       "2         C      V      L      non  \n",
       "3         T      M      L      non  \n",
       "4         G      K      E      non  \n",
       "..      ...    ...    ...      ...  \n",
       "83        A      F      I      non  \n",
       "84        G      A      A      syn  \n",
       "85        C      R      R      syn  \n",
       "86        C      G      G      syn  \n",
       "87        T      A      S      non  \n",
       "\n",
       "[88 rows x 8 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['contig',\n",
       " 'gene_num',\n",
       " 'pos',\n",
       " 'ref_base',\n",
       " 'alt_base',\n",
       " 'ref_aa',\n",
       " 'alt_aa',\n",
       " 'snp_type']"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.columns.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len( ndf )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing microcosm [35, 43, 51, 59, 67, 75, 83, 91]. . .\n",
      "\n",
      "Processing genome 1. . .\n",
      "\n",
      "Processing genome 2. . .\n",
      "\n",
      "Processing genome 3. . .\n",
      "\n",
      "Processing genome 4. . .\n",
      "\n",
      "Processing genome 5. . .\n",
      "\n",
      "Processing genome 6. . .\n",
      "\n",
      "Processing genome 7. . .\n",
      "\n",
      "Processing genome 9. . .\n",
      "\n",
      "Processing genome 10. . .\n",
      "\n",
      "Processing genome 11. . .\n",
      "\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-14-1f16337f1509>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     77\u001b[0m                 \u001b[0mtsnp\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m \u001b[0mrow\u001b[0m\u001b[0;34m[\u001b[0m \u001b[0;34m'contig'\u001b[0m \u001b[0;34m]\u001b[0m \u001b[0;34m)\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0;34m'_'\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m \u001b[0mrow\u001b[0m\u001b[0;34m[\u001b[0m \u001b[0;34m'gene_num'\u001b[0m \u001b[0;34m]\u001b[0m \u001b[0;34m)\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0;34m'_'\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m \u001b[0mrow\u001b[0m\u001b[0;34m[\u001b[0m \u001b[0;34m'pos'\u001b[0m \u001b[0;34m]\u001b[0m \u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     78\u001b[0m                 \u001b[0;32mif\u001b[0m \u001b[0mtsnp\u001b[0m \u001b[0;32min\u001b[0m \u001b[0msnp_detects\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 79\u001b[0;31m                     \u001b[0mndf\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mndf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m \u001b[0mdf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0miloc\u001b[0m\u001b[0;34m[\u001b[0m \u001b[0midx\u001b[0m \u001b[0;34m]\u001b[0m \u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     80\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m \u001b[0mndf\u001b[0m \u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     81\u001b[0m                 \u001b[0mndf\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mndf\u001b[0m\u001b[0;34m[\u001b[0m \u001b[0mdf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtolist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.linuxbrew/opt/python/lib/python3.7/site-packages/pandas/core/frame.py\u001b[0m in \u001b[0;36mappend\u001b[0;34m(self, other, ignore_index, verify_integrity, sort)\u001b[0m\n\u001b[1;32m   7116\u001b[0m                 \u001b[0mother\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mvalues\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreshape\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mother\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   7117\u001b[0m                 \u001b[0mindex\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 7118\u001b[0;31m                 \u001b[0mcolumns\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mcombined_columns\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   7119\u001b[0m             )\n\u001b[1;32m   7120\u001b[0m             \u001b[0mother\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mother\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_convert\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdatetime\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtimedelta\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.linuxbrew/opt/python/lib/python3.7/site-packages/pandas/core/frame.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, data, index, columns, dtype, copy)\u001b[0m\n\u001b[1;32m    438\u001b[0m                 \u001b[0mmgr\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0minit_dict\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m{\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mdata\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mindex\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcolumns\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdtype\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mdtype\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    439\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 440\u001b[0;31m                 \u001b[0mmgr\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0minit_ndarray\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mindex\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcolumns\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdtype\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mdtype\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcopy\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mcopy\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    441\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    442\u001b[0m         \u001b[0;31m# For data is list-like, or Iterable (will consume into list)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.linuxbrew/opt/python/lib/python3.7/site-packages/pandas/core/internals/construction.py\u001b[0m in \u001b[0;36minit_ndarray\u001b[0;34m(values, index, columns, dtype, copy)\u001b[0m\n\u001b[1;32m    202\u001b[0m             \u001b[0;31m# TODO: What about re-joining object columns?\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    203\u001b[0m             block_values = [\n\u001b[0;32m--> 204\u001b[0;31m                 \u001b[0mmake_block\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdvals_list\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mn\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mplacement\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mn\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mn\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mrange\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdvals_list\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    205\u001b[0m             ]\n\u001b[1;32m    206\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.linuxbrew/opt/python/lib/python3.7/site-packages/pandas/core/internals/construction.py\u001b[0m in \u001b[0;36m<listcomp>\u001b[0;34m(.0)\u001b[0m\n\u001b[1;32m    202\u001b[0m             \u001b[0;31m# TODO: What about re-joining object columns?\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    203\u001b[0m             block_values = [\n\u001b[0;32m--> 204\u001b[0;31m                 \u001b[0mmake_block\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdvals_list\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mn\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mplacement\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mn\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mn\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mrange\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdvals_list\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    205\u001b[0m             ]\n\u001b[1;32m    206\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.linuxbrew/opt/python/lib/python3.7/site-packages/pandas/core/internals/blocks.py\u001b[0m in \u001b[0;36mmake_block\u001b[0;34m(values, placement, klass, ndim, dtype, fastpath)\u001b[0m\n\u001b[1;32m   3275\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mklass\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3276\u001b[0m         \u001b[0mdtype\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdtype\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0mvalues\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdtype\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 3277\u001b[0;31m         \u001b[0mklass\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mget_block_type\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mvalues\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdtype\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   3278\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3279\u001b[0m     \u001b[0;32melif\u001b[0m \u001b[0mklass\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0mDatetimeTZBlock\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mis_datetime64tz_dtype\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mvalues\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.linuxbrew/opt/python/lib/python3.7/site-packages/pandas/core/internals/blocks.py\u001b[0m in \u001b[0;36mget_block_type\u001b[0;34m(values, dtype)\u001b[0m\n\u001b[1;32m   3234\u001b[0m         \u001b[0;32massert\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mis_datetime64tz_dtype\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mvalues\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3235\u001b[0m         \u001b[0mcls\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mDatetimeBlock\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 3236\u001b[0;31m     \u001b[0;32melif\u001b[0m \u001b[0mis_datetime64tz_dtype\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mvalues\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   3237\u001b[0m         \u001b[0mcls\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mDatetimeTZBlock\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3238\u001b[0m     \u001b[0;32melif\u001b[0m \u001b[0mis_interval_dtype\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdtype\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0mis_period_dtype\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdtype\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.linuxbrew/opt/python/lib/python3.7/site-packages/pandas/core/dtypes/common.py\u001b[0m in \u001b[0;36mis_datetime64tz_dtype\u001b[0;34m(arr_or_dtype)\u001b[0m\n\u001b[1;32m    572\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0marr_or_dtype\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    573\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0;32mFalse\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 574\u001b[0;31m     \u001b[0;32mreturn\u001b[0m \u001b[0mDatetimeTZDtype\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mis_dtype\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0marr_or_dtype\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    575\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    576\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.linuxbrew/opt/python/lib/python3.7/site-packages/pandas/core/dtypes/base.py\u001b[0m in \u001b[0;36mis_dtype\u001b[0;34m(cls, dtype)\u001b[0m\n\u001b[1;32m    254\u001b[0m         \u001b[0mdtype\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mgetattr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdtype\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"dtype\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdtype\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    255\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 256\u001b[0;31m         \u001b[0;32mif\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdtype\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mABCSeries\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mABCIndexClass\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mABCDataFrame\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdtype\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    257\u001b[0m             \u001b[0;31m# https://github.com/pandas-dev/pandas/issues/22960\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    258\u001b[0m             \u001b[0;31m# avoid passing data to `construct_from_string`. This could\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.linuxbrew/opt/python/lib/python3.7/site-packages/pandas/core/dtypes/generic.py\u001b[0m in \u001b[0;36m_check\u001b[0;34m(cls, inst)\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0;31m# objects\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      6\u001b[0m \u001b[0;32mdef\u001b[0m \u001b[0mcreate_pandas_abc_type\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mattr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcomp\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 7\u001b[0;31m     \u001b[0;34m@\u001b[0m\u001b[0mclassmethod\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      8\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_check\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcls\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0minst\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      9\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mgetattr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0minst\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mattr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"_typ\"\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mcomp\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "m01 = [ 35, 43, 51, 59, 67, 75, 83, 91 ]\n",
    "m02 = [ 36, 44, 52, 60, 68, 76, 84, 92 ]\n",
    "m03 = [ 37, 45, 53, 61, 69, 77, 85, 93 ]\n",
    "m04 = [ 38, 46, 54, 62, 70, 78, 86, 94 ]\n",
    "m05 = [ 39, 47, 55, 63, 71, 79, 87, 95 ]\n",
    "m06 = [ 40, 48, 56, 64, 72, 80, 88, 96 ]\n",
    "m07 = [ 41, 49, 57, 65, 73, 81, 89, 97 ]\n",
    "m08 = [ 42, 50, 58, 66, 74, 82, 90, 98 ]\n",
    "m09 = [ 99, 101, 103, 105, 107, 109, 111, 113 ]\n",
    "m10 = [ 100, 102, 104, 106, 108, 110, 112, 114 ]\n",
    "\n",
    "allms = [ m01, m02, m03, m04, m05, m06, m07, m08, m09, m10 ]\n",
    "g_num_list = list( range( 1, 25 ) ) + list( range( 26, 35 ) )\n",
    "\n",
    "for tm00 in allms:\n",
    "    print( 'Processing microcosm ' + str( tm00 ) + '. . .\\n' )\n",
    "    for g_num in g_num_list:\n",
    "        print( 'Processing genome ' + str( g_num ) + '. . .\\n' )\n",
    "        snp_detects = {}\n",
    "        is_invalid = False\n",
    "        for this_idx, this_time in enumerate( tm00 ):\n",
    "            # Going from time point to time point.\n",
    "            try:\n",
    "                df = pd.read_csv( '/home/akshitg/lte/metagenomes_lte/freebayes_out/snps/fb_processed/mg_' \n",
    "                                  + str( this_time ) + '_gnm_' + str( g_num ) + \n",
    "                                  '_processed_snps.tsv', sep='\\t', index_col=0 )\n",
    "                \n",
    "                if len( df ) == 0:\n",
    "                    is_invalid = True\n",
    "                    continue\n",
    "            except:\n",
    "                is_invalid = True\n",
    "                continue\n",
    "\n",
    "            # Now go SNP by SNP.\n",
    "            for idx, row in  df.iterrows():\n",
    "                tsnp = str( row[ 'contig' ] ) + '_' + str( row[ 'gene_num' ] ) + '_' + str( row[ 'pos' ] )\n",
    "                if tsnp in snp_detects:\n",
    "                    snp_detects[ tsnp ] += 1\n",
    "                else:\n",
    "                    snp_detects[ tsnp ] = 1\n",
    "\n",
    "        if is_invalid:\n",
    "            continue\n",
    "                        \n",
    "        THRESHOLD_N = 3\n",
    "        bad_snps = []\n",
    "        for tsnp in snp_detects:\n",
    "            if snp_detects[ tsnp ] < THRESHOLD_N:\n",
    "                bad_snps.append( tsnp )\n",
    "\n",
    "        for tsnp in bad_snps:\n",
    "            snp_detects.pop( tsnp, None )\n",
    "\n",
    "\n",
    "        for this_idx, this_time in enumerate( tm00 ):\n",
    "            try:\n",
    "                # Going from time point to time point.\n",
    "                df = pd.read_csv( '/home/akshitg/lte/metagenomes_lte/freebayes_out/snps/fb_processed/mg_' \n",
    "                                  + str( this_time ) + '_gnm_' + str( g_num ) + \n",
    "                                  '_processed_snps.tsv', sep='\\t', index_col=0 )\n",
    "                \n",
    "                if len( df ) == 0:\n",
    "                    is_invalid = True\n",
    "                    continue\n",
    "            except:\n",
    "                is_invalid = True\n",
    "                continue\n",
    "\n",
    "            ndf = pd.DataFrame()\n",
    "\n",
    "            # Now go SNP by SNP.\n",
    "            for idx, row in  df.iterrows():\n",
    "                tsnp = str( row[ 'contig' ] ) + '_' + str( row[ 'gene_num' ] ) + '_' + str( row[ 'pos' ] )\n",
    "                if tsnp in snp_detects:\n",
    "                    ndf = ndf.append( df.iloc[ idx ] )\n",
    "            if len( ndf ):\n",
    "                ndf = ndf[ df.columns.tolist() ]\n",
    "                ndf.to_csv( '/home/akshitg/lte/metagenomes_lte/freebayes_out/snps/fb_final/mg_' \n",
    "                            + str( this_time ) + '_gnm_' + str( g_num ) + '_final_snps.tsv', sep='\\t' )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
